contracts/core/src/token/OmniGasPump.sol

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity 0.8.24;
+
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { PausableUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/PausableUpgradeable.sol";
+import { XAppUpgradeable } from "src/pkg/XAppUpgradeable.sol";
+import { FeeOracleV1 } from "src/xchain/FeeOracleV1.sol";
+import { ConfLevel } from "src/libraries/ConfLevel.sol";
+import { OmniXFund } from "./OmniXFund.sol";
+
+/**
+ * @title OmniGasPump
+ * @notice A unidirectional cross-chain gas exchange, allowing users to swap native ETH for native OMNI.[]
+ */
+contract OmniGasPump is XAppUpgradeable, OwnableUpgradeable, PausableUpgradeable {
+    event MaxSwapSet(uint256 max);
+    event FundSet(address fund);
+    event PctCutSet(uint256 pct);
+    event Refilled(address indexed recipient, uint256 paid, uint256 fee, uint256 cut, uint256 amtOMNI);
+
+    /// @notice Gas limit passed to OmniXFund.settleUp xcall
+    uint64 internal constant SETTLE_GAS = 100_000;
+
+    /// @notice Denominator for pct cut
+    uint256 internal constant PCT_DENOM = 1000;
+
+    /// @notice Address of OmniXFund on Omni
+    address public fund;
+
+    /// @notice Max amt (in native token) that can be swapped in a single tx
+    uint256 public maxSwap;
+
+    /// @notice Percentage cut (over PCT_DENOM) taken by this contract for each swap
+    ///     Used to disencentivize spamming
+    uint256 public pctCut;
+
+    /// @notice Map address to total historical fillUps(), denominated in OMNI.
+    mapping(address => uint256) public bought;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    struct InitParams {
+        address fund;
+        address portal;
+        address owner;
+        uint256 maxSwap;
+        uint256 pctCut;
+    }
+
+    function initialize(InitParams calldata p) external initializer {
```

Comment:
I think we should probably keep global params (i.e. pctCut) on Omni, rather than on rollups. Global state! Only keep parameters in 1 place if possible

Max swap makes sense to keep here ofc, and the others are required. But no need to track pct cut in many places

Reply:
Not a bad call. 

Thought part of the design is that this `OmniXFund` can be used for any app we have that ends up funding people with OMNI. OmniGasPump would just be one of them. We can add more by just `auth`ing new contracts

But then again, it's hard to build for the future. Sometimes better to just solve the problem at hand and nothing more

Reply:
Sorry did mention that this was a design goal anywhere ^

Reply:
Also doing all the cuts before coming to a total "owed" value on the source chain is useful for the  accounting we do

Reply:
Ehh I think we should not try to optimize for something we don't know about. You can make the call here, maybe you have something in mind that I don't. My only request is that you consider maintenance simplicity, I have not thought through what is necessary in each case, so will leave that to you

Reply:
If we do make this "not a generic fund" we can rename it to OmniGasStation which I've been wanting to do

Reply:
Wrapping this up. (copying from slack for posterity)

Tried moving toll to the singleton contract on Omni. Unfortunately it does make the book keeping trickier.

Because we're working with "total" amounts - total owed vs total settled - moving the toll to the gas station means that changing the toll changes the toll for all historical swaps.

Reply:
TLDR: going to keep toll on gas pumps